File: backend-dotnet/Infrastructure/Services/SearchService.cs
using Application.Common;
using Application.Interfaces;
using Domain.Enums;
using Microsoft.Extensions.Logging;

namespace Infrastructure.Services
{
    public class SearchService : ISearchService
    {
        private readonly IPostRepository _postRepository;
        private readonly IUserFollowRepository _userFollowRepository;
        private readonly IPostTagRepository _postTagRepository;
        private readonly IElasticService _elastic;
        private readonly ICacheService _cacheService;
        private readonly ILogger<SearchService> _logger;

        public SearchService(
            IPostRepository postRepository,
            IUserFollowRepository userFollowRepository,
            IPostTagRepository postTagRepository,
            IElasticService elastic,
            ICacheService cacheService,
            ILogger<SearchService> logger)
        {
            _postRepository = postRepository;
            _userFollowRepository = userFollowRepository;
            _postTagRepository = postTagRepository;
            _elastic = elastic;
            _cacheService = cacheService;
            _logger = logger;
        }

        public async Task<object> SearchAsync(string query, string? userIdStr)
        {
            Guid? userId = Guid.TryParse(userIdStr, out var parsedUserId) ? parsedUserId : null;
            List<Guid> followingIds = [];

            if (userId != null)
            {
                followingIds = await _userFollowRepository.GetFollowingIdsAsync(userId.Value);
            }

            if (query.StartsWith('#'))
            {
                var tag = query.TrimStart('#').ToLowerInvariant();

                var postIds = await _postTagRepository.GetPostIdsByTagAsync(tag);
                var posts = await _postRepository.GetPostsByIdsAsync(postIds);

                var filtered = posts.Where(p => CanViewPost(p.Visibility, p.UserId, userId, followingIds));

                return new
                {
                    query,
                    users = Array.Empty<object>(),
                    hashtags = new[] { tag },
                    posts = filtered.Select(p => new
                    {
                        p.Id,
                        p.Caption,
                        p.ImageUrl,
                        p.CreatedAt,
                        likes = p.Likes.Count,
                        liked = userId != null && p.Likes.Any(l => l.UserId == userId),
                        user = new
                        {
                            p.User.Id,
                            p.User.Username,
                            p.User.ProfileImageUrl
                        }
                    })
                };
            }
            else
            {
                var posts = await _elastic.SearchPostsAsync(query);
                var users = await _elastic.SearchUsersAsync(query);
                var tags = await _elastic.SearchTagsAsync(query.ToLowerInvariant());

                var filtered = posts.Where(p => CanViewPost(p.Visibility, p.UserId, userId, followingIds));

                return new
                {
                    query,
                    hashtags = tags.Select(t => t.Name),
                    users = users.Select(u => new
                    {
                        u.Id,
                        u.Username,
                        u.ProfileImageUrl
                    }),
                    posts = filtered.Select(p => new
                    {
                        p.Id,
                        p.Caption,
                        p.ImageUrl,
                        p.CreatedAt,
                        likes = 0,
                        liked = false,
                        user = new
                        {
                            Id = p.UserId,
                            p.Username
                        }
                    })
                };
            }
        }

        public async Task<object> AutocompleteAsync(string query)
        {
            var normalizedQuery = query.Trim().ToLower();
            var cacheKey = $"autocomplete:{normalizedQuery}";

            var cached = await _cacheService.GetAsync<object>(cacheKey);
            if (cached != null)
            {
                _logger.LogDebug("Autocomplete cache hit for query: {Query}", query);
                return cached;
            }

            _logger.LogDebug("Autocomplete cache miss for query: {Query}", query);

            var result = await GetAutocompleteFromElasticsearch(query);

            await _cacheService.SetAsync(cacheKey, result, TimeSpan.FromMinutes(5));

            _logger.LogInformation("Autocomplete result cached for query: {Query}", query);
            return result;
        }

        private async Task<object> GetAutocompleteFromElasticsearch(string query)
        {
            query = query.Trim();

            if (query.StartsWith('#'))
            {
                var keyword = query.TrimStart('#').ToLowerInvariant();
                var tags = await _elastic.SearchTagsAsync(keyword, 3);
                var users = await _elastic.SearchUsersAsync(keyword, 10);

                if (tags.Count != 0)
                {
                    return new
                    {
                        query,
                        mode = "tag",
                        posts = Array.Empty<object>(),
                        hashtags = tags.Select(t => t.Name),
                        tags = tags.Select(t => t.Name),
                        users = users.Select(u => new
                        {
                            u.Id,
                            u.Username,
                            u.ProfileImageUrl
                        }),
                        cached_at = DateTime.UtcNow
                    };
                }
                else
                {
                    return new
                    {
                        query,
                        mode = "tag",
                        posts = Array.Empty<object>(),
                        hashtags = Array.Empty<string>(),
                        tags = Array.Empty<string>(),
                        users = Array.Empty<string>(),
                        fallback = $"Search for \"{query}\"",
                        cached_at = DateTime.UtcNow
                    };
                }
            }

            var tagResults = await _elastic.SearchTagsAsync(query.ToLowerInvariant(), 3);
            var userResults = await _elastic.SearchUsersAsync(query, 10);

            return new
            {
                query,
                mode = "default",
                posts = Array.Empty<object>(),
                hashtags = tagResults.Select(t => t.Name),
                tags = tagResults.Select(t => t.Name),
                users = userResults.Select(u => new
                {
                    u.Id,
                    u.Username,
                    u.ProfileImageUrl
                }),
                cached_at = DateTime.UtcNow
            };
        }

        private static bool CanViewPost(
            PostVisibility visibility,
            Guid ownerId,
            Guid? currentUserId,
            IReadOnlyCollection<Guid> followingIds)
        {
            if (visibility == PostVisibility.Public)
                return true;

            if (!currentUserId.HasValue)
                return false;

            if (ownerId == currentUserId.Value)
                return true;

            return visibility == PostVisibility.FollowersOnly && followingIds.Contains(ownerId);
        }
    }
}

File: backend-java/src/main/java/com/jerrygram/infrastructure/services/SearchService.java
package com.jerrygram.infrastructure.services;

import com.jerrygram.application.dtos.PostDto;
import com.jerrygram.application.dtos.SearchResultDto;
import com.jerrygram.application.dtos.UserProfileDto;
import com.jerrygram.application.interfaces.ICacheService;
import com.jerrygram.application.interfaces.ISearchService;
import com.jerrygram.domain.entities.Post;
import com.jerrygram.domain.entities.User;
import com.jerrygram.domain.enums.PostVisibility;
import com.jerrygram.infrastructure.repositories.PostRepository;
import com.jerrygram.infrastructure.repositories.PostTagRepository;
import com.jerrygram.infrastructure.repositories.UserFollowRepository;
import com.jerrygram.infrastructure.repositories.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;

import java.time.Duration;
import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Slf4j
public class SearchService implements ISearchService {

    private final UserRepository userRepository;
    private final PostRepository postRepository;
    private final UserFollowRepository userFollowRepository;
    private final PostTagRepository postTagRepository;
    private final ICacheService cacheService;

    @Override
    public SearchResultDto search(String query, String userId) {
        log.info("Searching for: {} by user: {}", query, userId);

        UUID parsedUserId = parseUserId(userId);
        List<UUID> followingIds = getFollowingIds(parsedUserId);
        
        if (query.startsWith("#")) {
            return searchByHashtag(query, parsedUserId, followingIds);
        } else {
            return searchGeneral(query, parsedUserId, followingIds);
        }
    }
    
    private UUID parseUserId(String userIdStr) {
        if (userIdStr == null || userIdStr.isEmpty()) {
            return null;
        }
        try {
            return UUID.fromString(userIdStr);
        } catch (IllegalArgumentException e) {
            log.warn("Invalid userId format: {}", userIdStr);
            return null;
        }
    }
    
    private List<UUID> getFollowingIds(UUID userId) {
        if (userId == null) {
            return List.of();
        }
        return userFollowRepository.getFollowingIds(userId);
    }
    
    private SearchResultDto searchByHashtag(String query, UUID userId, List<UUID> followingIds) {
        String tag = query.substring(1).toLowerCase();
        var postIds = postTagRepository.getPostIdsByTag(tag);
        var posts = postRepository.findByIdIn(postIds);
        
        var filteredPosts = posts.stream()
                .filter(post -> isPostVisible(post, userId, followingIds))
                .map(this::mapToPostDto)
                .toList();
        
        return SearchResultDto.builder()
                .users(List.of())
                .posts(filteredPosts)
                .hashtags(List.of(tag))
                .build();
    }
    
    private SearchResultDto searchGeneral(String query, UUID userId, List<UUID> followingIds) {
        var pageable = PageRequest.of(0, 10);
        
        var posts = postRepository.findByCaptionContainingIgnoreCase(query, pageable)
                .stream()
                .filter(post -> isPostVisible(post, userId, followingIds))
                .map(this::mapToPostDto)
                .toList();
        
        var users = userRepository.findByUsernameContainingIgnoreCase(query, pageable)
                .stream()
                .map(this::mapToUserDto)
                .toList();
        
        return SearchResultDto.builder()
                .users(users)
                .posts(posts)
                .hashtags(List.of())
                .build();
    }
    
    private boolean isPostVisible(Post post, UUID userId, List<UUID> followingIds) {
        return post.getVisibility() == PostVisibility.Public ||
               (post.getVisibility() == PostVisibility.FollowersOnly && 
                userId != null && followingIds.contains(post.getUser().getId()));
    }
    
    private PostDto mapToPostDto(Post post) {
        return PostDto.builder()
                .id(post.getId())
                .caption(post.getCaption())
                .imageUrl(post.getImageUrl())
                .visibility(post.getVisibility())
                .author(mapToUserDto(post.getUser()))
                .likesCount(post.getLikesCount())
                .commentsCount(post.getCommentsCount())
                .createdAt(post.getCreatedAt())
                .build();
    }
    
    private UserProfileDto mapToUserDto(User user) {
        return UserProfileDto.builder()
                .id(user.getId())
                .username(user.getUsername())
                .email(user.getEmail())
                .profileImageUrl(user.getProfileImageUrl())
                .createdAt(user.getCreatedAt())
                .followers(0)
                .followings(0)
                .build();
    }

    @Override
    public SearchResultDto autocomplete(String query) {
        log.info("Autocomplete for: {}", query);
        
        String cacheKey = "autocomplete:" + query.trim().toLowerCase();
        
        // Try cache first
        var cached = tryGetFromCache(cacheKey);
        if (cached != null) {
            log.debug("Autocomplete cache hit for query: {}", query);
            return cached;
        }
        
        log.debug("Autocomplete cache miss for query: {}", query);
        
        SearchResultDto result = performAutocomplete(query.trim());
        
        // Cache for 5 minutes
        trySetCache(cacheKey, result);
        
        return result;
    }
    
    private SearchResultDto tryGetFromCache(String cacheKey) {
        try {
            var cached = cacheService.get(cacheKey, SearchResultDto.class);
            return cached.orElse(null);
        } catch (Exception e) {
            log.warn("Cache get failed for key: {}", cacheKey, e);
            return null;
        }
    }
    
    private void trySetCache(String cacheKey, SearchResultDto result) {
        try {
            cacheService.set(cacheKey, result, Duration.ofMinutes(5));
            log.info("Autocomplete result cached");
        } catch (Exception e) {
            log.warn("Cache set failed for key: {}", cacheKey, e);
        }
    }
    
    private SearchResultDto performAutocomplete(String query) {
        if (query.startsWith("#")) {
            return autocompleteHashtag(query.substring(1).toLowerCase());
        } else {
            return autocompleteGeneral(query);
        }
    }
    
    private SearchResultDto autocompleteHashtag(String keyword) {
        var hashtags = postTagRepository.findHashtagsByPrefix(keyword)
                .stream()
                .limit(3)
                .toList();
        
        var users = userRepository.findByUsernameStartingWithIgnoreCase(keyword, PageRequest.of(0, 10))
                .stream()
                .map(this::mapToUserDto)
                .toList();
        
        return SearchResultDto.builder()
                .users(users)
                .posts(List.of())
                .hashtags(hashtags)
                .build();
    }
    
    private SearchResultDto autocompleteGeneral(String query) {
        var hashtags = postTagRepository.findHashtagsByPrefix(query.toLowerCase())
                .stream()
                .limit(3)
                .toList();
        
        var users = userRepository.findByUsernameStartingWithIgnoreCase(query, PageRequest.of(0, 10))
                .stream()
                .map(this::mapToUserDto)
                .toList();
        
        return SearchResultDto.builder()
                .users(users)
                .posts(List.of())
                .hashtags(hashtags)
                .build();
    }
}